decentraland/unity-renderer
# Feat: Embed Base Wearables

## What does this PR change?

Closes #3872 
Closes #3919

### Loading modes
Now there is an abstraction to define how assets can be retrieved, and it's configured via IoC (`ServiceFactory`):
```
result.Register<ITextureAssetResolver>(() => new TextureAssetResolver(new Dictionary<AssetSource, ITextureAssetProvider>
            {
                { AssetSource.EMBEDDED, new EmbeddedTextureProvider() },
                { AssetSource.WEB, new AssetTextureWebLoader() }
            }));

            result.Register<IAssetBundleResolver>(() => new AssetBundleResolver(new Dictionary<AssetSource, IAssetBundleProvider>
            {
                { AssetSource.EMBEDDED, new EmbeddedAssetBundleProvider() },
                { AssetSource.WEB, new AssetBundleWebLoader(DataStore.i.featureFlags, DataStore.i.performance) }
            }, new EditorAssetBundleProvider()));
```
 `AssetPromise_Texture` and `AssetPromise_AB` accept `AssetSource` as a flag of permitted source:
- `AssetPromise_Texture` has `WEB` by default, to avoid an additional overhead only [wearables] thumbnails pass `ALL`
- `AssetPromise_AB` has `ALL` by default


Adding modes is as easy as adding an enum according to its order and implementing specific providers. 

### Providers and Resolvers Layer
Refer to `IAssetBundleResolver` and `ITextureAssetResolver`
This layer is up to consuming providers and trying to fetch assets from them one by one.
It can be easily extended in application to other types of assets according to the same rule set.

Refer to `IAssetBundleProvider` and `ITextureAssetProvider`
There are 2 main implementations at the moment: `Web` and `Embedded`. The latter aims for downloading Resources from the build itself before trying to make any web request.
It's up to the implementation to decide on how to act: it can be downloading from `Resources`, from `StreamingAssets`, directly from `Url`, or a dynamic cache: whatever is needed.

**After the tests I disabled Embedded Asset Bundles for `WebGL` as I saw no benefits it can grant:**
- `StreamingAssets` are not included in the binary itself (and ABs should not)
- They are just accessed by a Web URL, and it's not under a balancer
- But we already have a URL to access them so it would just duplicate the storage
- Moreover, we would need still to enable caching in `IndexedDB` (locally) to benefit somehow from it so it would result in having 3 copies in different places

### Embedment logic (Editor scripts)
Refer to `AssetEmbedder` and sub-classes
Currently it's used for `Wearables` collections only but can be used for any assets.
It just downloads the assets from the given end-points and stores them on the drive accordingly.
`Thumbnails` are stored in `Resources`.
`AssetBundles` are stored in `StreamingAssets`

### AB Caching

- Flag `AB_CACHING` introduced
- The logic is encapsulated in `AssetBundleWebLoader`
- `DCL_CachedXHR` is removed as it was disabled anyway and didn't work
- Unity uses `IndexedDB` for WebGL to cache ABs

### Refactoring
In order to overcome obsolescence, and make code reusable and robust the following concepts were modified:

- `AssetPromise_Texture` and `AssetPromise_AB` to work with `UniTasks` and asynchronous `Resolvers` instead of hard-coded logic. The adjacent layers were not touched, just a little adapted if needed (it's a giant chunk of legacy)
- `AssetBundleDepMapLoadHelper` to provide `async` API to resolve ABs' dependencies and be reusable for both Editor and Runtime code.

## How to test the changes?

- Make sure to use `ENABLE_AB_CACHING` feature flag
- Base Wearables' thumbnails are embedded and loaded faster
- Asset Bundles are cached after the first retrieval, keep in mind not everything can be kept in the cache so Unity handles it smartly.
- You should validate it:
  - Enable `enable_asset_resolver_verbose` feature flag
  - Observe JS dev console logs:
     - If the asset is cached you will see smth like `Asset Bundle bafybeibgpkdsthygnoqj4t4aqyy7mtobqxqwlceish7gkqvxogiin7dpmm is cached: True`, otherwise `... is cached: False`
     - If the texture is embedded you will see `Texture https://peer.melonwave.com/content/contents/QmY8nLXpBU1VtYyf4RPEHpoAievaXc1iBcUCNR91MwfhXw loaded from DCL.EmbeddedTextureProvider`, otherwise `Texture https://events.decentraland.org/poster/bccb125f639b8878.png loaded from DCL.AssetTextureWebLoader`
     - Without `enable_asset_resolver_verbose` no such logs should be observed at all, they are for debugging purposes only
- Asset Bundles and textures should work just like before (travel in the world, no unexpected behavior should be observed)

1. Go to: https://play.decentraland.zone/?renderer-branch=feat/embed_base_wearables&ENABLE_AB_CACHING
2. ...

## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md
